portfolio/src/main/java/com/google/sps/servlets/GameDataServlet.java

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.gameChoice;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+@WebServlet("/GameDataServlet")
+public class GameDataServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+    throws IOException {
+    Query query = new Query("gameChoice");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    List<gameChoice> games = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String game = (String) entity.getProperty("game");
+      long count = (long) entity.getProperty("count");
+
+      gameChoice gameVote = new gameChoice(game, count);
+      games.add(gameVote);
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJson(games));
+  }
+
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+    throws IOException {
+    String game = request.getParameter("game");
+    Entity gameEntity = new Entity("gameChoice");
+    if (!checkDuplicate(game)) {
+      gameEntity.setProperty("game", game);
+      gameEntity.setProperty("count", 1);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(gameEntity);
+    }
+
+    response.sendRedirect("/chart.html");
+  }
+
+  private String convertToJson(List<gameChoice> games) {
+    String json = "{";
```

Comment:
You've probably already seen that converting a Java object to JSON is pretty tedious this way. You almost always want to use a library that knows how to write/read JSON data instead. Here's one example: https://mkyong.com/java/jackson-2-convert-java-object-to-from-json/.

Let's leave this as is for now to let you focus on this week's assignments, but add a TODO comment to use a JSON library in the future.

Reply:
I tried using Gson like last time but gson would do something like this `[ {"game" : "LEAGUE OF LEGENDS" , "count" : 6} , {"game" : "WORLD OF WARCRAFT" , "count" : 1} ]  ` . This would usually be fine but google charts expects the json in this form `{"LEAGUE OF LEGENDS" : 6 , "WORLD OF WARCRAFT" : 1}`, so I decided to just manually construct the json the way google charts expects so I would not waste time configuring gson.

Reply:
Oh, I see. If you need to change the format slightly, often you'll want to create a Java HashMap from your Java Object then pass that Map to the JSON library. It'll end up looking much nicer than what you currently have.